Tests: Add new `assertEqualHTML` assertion

ockham opened the pull request:
## Proposed changes

Add `assertEqualHTML` method to `WP_UnitTestClass` for tests comparing HTML (potentially including block markup).

Comparing HTML markup can be error prone and fragile. Typically we're interested in _semantically equivalent_ HTML, which is what this assertion allows for; i.e. it has built-in tolerance for different attribute and class name order.

Internally, it builds a deterministic tree string representation of the HTML (using the HTML API) and compares the results.

The format of the tree is inspired by the [HTML5lib-tests tree format.](https://github.com/html5lib/html5lib-tests/blob/a9f44960a9fedf265093d22b2aa3c7ca123727b9/tree-construction/README.md). It is extended with a special representation of block delimiters and their attributes.

This format also makes it easier to visually spot the differences between the two strings of HTML markup, as reported by the assertion. Quoting the inline documentation (PHPDoc) that's part of this PR:

```
For example, consider the following block markup:

    <!-- wp:separator {"className":"is-style-default has-custom-classname","style":{"spacing":{"margin":{"top":"50px","bottom":"50px"}}},"backgroundColor":"accent-1"} -->
        <hr class="wp-block-separator is-style-default has-custom-classname" style="margin-top: 50px; margin-bottom: 50px" />
    <!-- /wp:separator -->

This will be represented as:

    BLOCK["core/separator"]
      {
        "backgroundColor": "accent-1",
        "className": "has-custom-classname is-style-default",
        "style": {
          "spacing": {
            "margin": {
              "top": "50px",
              "bottom": "50px"
            }
          }
        }
      }
      <hr>
        class="has-custom-classname is-style-default wp-block-separator"
        style="margin-top:50px;margin-bottom:50px;"
```

## Note

_Edit: Fixed by @sirreal, see https://github.com/WordPress/wordpress-develop/pull/8882#issuecomment-2945209149._

~This PR adds the new `assertEqualBlockMarkup` method to the `WP_UnitTestCase` class. Note that `class Tests_Dependencies_Scripts`, which is a subclass of `WP_UnitTestCase`, already contains a `protected` method named `assertEqualMarkup`. The latter is exclusively used for dependencies/scripts tests. The new `assertEqualBlockMarkup` cannot be used as a drop-in replacement, due to some differences in what it considers semantically equivalent HTML.~

~The existence of that protected `assertEqualMarkup` method was a consideration in naming the new method `assertEqualBlockMarkup`, so that it wouldn't collide. It seems like a reasonable choice, given that `assertEqualBlockMarkup` is indeed capable not only of parsing HTML, but also _block_ markup.~

~In a future iteration, we might choose to retire the existing `assertEqualMarkup` method (and use `assertEqualBlockMarkup` for dependencies/scripts tests). This would then also pave the way for a general-purpose `assertEqualMarkup` method (which is able to parse HTML but agnostic of block markup), should the need arise.~

Props @sirreal @dmsnell 
Trac ticket: https://core.trac.wordpress.org/ticket/63527

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

ockham committed:
Tests: Add assertEqualMarkup trait

ockham committed:
Add test coverage for assertion helper method

ockham committed:
Fix include path

ockham committed:
Remove now-obsolete method of the same name from Tests_Dependencies_Scripts

ockham committed:
Remove obsolete require_lib

ockham committed:
Add since PHPDoc

ockham committed:
Move assertion to class WP_UnitTestCase_Base

ockham committed:
Get rid of trait

ockham committed:
WPCS

ockham committed:
Remove potentially obsolete PHPCS ignore

ockham commented on tests/phpunit/tests/build-tree-representation.php:
```diff
@@ -0,0 +1,114 @@
+		$this->expectException( 'PHPUnit\Framework\ExpectationFailedException' );
+		$this->assertEqualMarkup( $expected, $actual );
+	}
+}
```
These tests could be potentially absorbed into the existing [`Test_Utils` class](https://github.com/WordPress/wordpress-develop/blob/f2cf6a3f2497760bf52c6950eaa3f7f582ae40c6/tests/phpunit/tests/utils.php) (which is the only other test with `@group testsuite`.)

ockham commented on tests/phpunit/includes/build-tree-representation.php:
```diff
@@ -0,0 +1,264 @@
+ * @param string      $html             Given test HTML.
+ * @return string|null Tree structure of parsed HTML, if supported, else null.
+ */
+function build_tree_representation( string $html, ?string $fragment_context ) {
```
This function could be potentially moved into `tests/phpunit/includes/utils.php`.

dmsnell replied:
also its name is widely general and could benefit from narrowing. perhaps something like `html_debug_as_tree_string()` or even just `html_debug_as_tree()` or `build_equivalent_html_semantic_tree()`?

ockham replied:
True! I like the last suggestion the most: e8ceff2d40

ockham committed:
Revert "Remove now-obsolete method of the same name from Tests_Dependencies_Scripts"

This reverts commit c01a3c14cea7b3c8bc7ed550247e25e3c8843d6e.

ockham committed:
Bring back assertEqualMarkup override, harmonize signature

ockham committed:
Remove trailing newline

ockham committed:
Fix class name

ockham committed:
Tweak tests

ockham committed:
Remove strict_types declaration

ockham committed:
Fix heredoc indentation for PHP 7.2 compat :rolleyes:

sirreal commented on tests/phpunit/includes/build-tree-representation.php:
```diff
@@ -0,0 +1,263 @@
+/**
+ * Generates the tree-like structure represented in the Html5lib tests.
+ *
+ * @see https://github.com/html5lib/html5lib-tests/blob/a9f44960a9fedf265093d22b2aa3c7ca123727b9/tree-construction/README.md
```
This is committed here for the HTML5lib test suite so it may be better to reference a link that's internal to this repo: https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/data/html5lib-tests/tree-construction/README.md

It may be good to explain that the tree representation is extended with special representation for WordPress blocks and their attributes.

dmsnell replied:
yeah some examples here would help, even though this itself is test code. additionally, if we are adding block semantics and class-name rearrangement, we probably ought to showcase and explain those.

ockham replied:
Good points! Changed reference in https://github.com/WordPress/wordpress-develop/pull/8882/commits/7cf7c64edfaee4c98d145687840101ebf30fd549, and added more commentary and an example in d0d46feae6.

sirreal commented on tests/phpunit/tests/dependencies/scripts.php:
```diff
@@ -3017,11 +3017,16 @@ protected function parse_markup_fragment( $markup ) {
+	 * @param string      $message          Message.
 	 */
-	protected function assertEqualMarkup( $expected, $actual, $message = '' ) {
+	public function assertEqualMarkup( string $expected, string $actual, ?string $fragment_context = '<body>', $message = 'HTML markup was not equivalent.' ): void {
```
Did you try having this be a wrapper that called `parent::assertEqualMarkup`?

ockham replied:
Do you mean to rename it so they can be "properly" distinct (including distinct function signatures)?

ockham replied:
FWIW, I tried removing the subclass' `assertEqualMarkup` (which predates this PR), but it's semantically different from the "new" one -- I got around 50 test failures for the existing script dependencies tests. I didn't want to block this PR by sorting them all out.

ockham replied:
I ended up reverting the changes to this method, instead option to rename the newly introduced assertion to `assertEqualBlockMarkup()`.

dmsnell commented on tests/phpunit/includes/build-tree-representation.php:
```diff
@@ -0,0 +1,263 @@
+
+	if ( $processor->paused_at_incomplete_token() ) {
+		throw new Error( 'Paused at incomplete token.' );
+	}
```
while it would probably rearrange things in ways we don’t want, for the PHP 8.4+ environments we can also convert via `DOM\HtmlDocument` and re-run, knowing that we are able to parse every well-formed and valid HTML5 document in the HTML API.

In some explorations I have done this. While it’s not great for production code, because lots of inbound HTML contains things like so-called “nested” `A` elements, and that leads to double-parsing on every request, it works fine for tests.

```php
if ( $did_not_parse && class_exists( 'DOM\HtmlDocument' ) ) {
	$dom = DOM\HtmlDocument::createFromString( $html );
	return build_tree_representation( $dom->saveHtml() );
}
```

of course this doesn’t take into consideration the fragment context. given that the tree representation here only accepts one input HTML, it’s possible this would be better handled in calling code, e.g. in `assertEqualMarkup` where it would test if the tree-building failed, and if so, normalize _both_ the actual and expected through `DOM\HtmlDocument` and re-running.

of course in pre-8.4 environments we don’t have this ability but as long as the tests run _at least_ in 8.4 and pass we can be confident in the results.

ockham replied:
Done in 5ff8ad4c891f6eaf10d0c2bfffe7c2e768232e95.

dmsnell commented on tests/phpunit/includes/build-tree-representation.php:
```diff
@@ -0,0 +1,263 @@
+				// Comments must be "<" then "!-- " then the data then " -->".
+				$comment = "<!--{$processor->get_full_comment_text()}-->";
+
+				// Maybe the comment is a block delimiter.
```
okay this definitely looks like a good case for `scan_delimiters()`, though I did propose at one point exposing the PCRE pattern for finding these, and that could also be helpful here since it grabs opening, closing, and void delimiters in one go

dmsnell replied:
I don’t think it would be terrible here to borrow the PCRE pattern in the meantime from the default block parser.

also, it looks like you can actually take this from `WP_Block_Parser` directly

```php
$parser           = new WP_Block_Parser();
$parser->document = $comment;
$parser->offset   = 0;
list( $delimiter_type, $block_name, $attrs, $start_offset, $token_length ) = $parser->next_token();

switch ( $delimiter_type ) {
	case 'no-more-tokens':
	case 'void-block':
	case 'block-closer':
	case 'block-opener':
}
```

dmsnell replied:
I don’t think it would be terrible here to borrow the PCRE pattern in the meantime from the default block parser.

```php
/*
 * aye the magic
 * we're using a single RegExp to tokenize the block comment delimiters
 * we're also using a trick here because the only difference between a
 * block opener and a block closer is the leading `/` before `wp:` (and
 * a closer has no attributes). we can trap them both and process the
 * match back in PHP to see which one it was.
 */
$has_match = preg_match(
	'/<!--\s+(?P<closer>\/)?wp:(?P<namespace>[a-z][a-z0-9_-]*\/)?(?P<name>[a-z][a-z0-9_-]*)\s+(?P<attrs>{(?:(?:[^}]+|}+(?=})|(?!}\s+\/?-->).)*+)?}\s+)?(?P<void>\/)?-->/s',
	$this->document,
	$matches,
	PREG_OFFSET_CAPTURE,
	$this->offset
);
```

also, it looks like you can actually take this from `WP_Block_Parser` directly

```php
$parser           = new WP_Block_Parser();
$parser->document = $comment;
$parser->offset   = 0;
list( $delimiter_type, $block_name, $attrs, $start_offset, $token_length ) = $parser->next_token();

switch ( $delimiter_type ) {
	case 'no-more-tokens':
	case 'void-block':
	case 'block-closer':
	case 'block-opener':
}
```

ockham replied:
Nice! Done in https://github.com/WordPress/wordpress-develop/pull/8882/commits/3fbf857cf1fdb401ba7ea22fde71b02c1e624fc8, with a bit of polish in subsequent commits.

dmsnell commented on tests/phpunit/includes/build-tree-representation.php:
```diff
@@ -0,0 +1,263 @@
+					if ( isset( $block_attrs['className'] ) ) {
+						// Normalize class name order, as we need to be tolerant of different orders.
+						// (Style attributes don't need this treatment, as they are parsed into a nested array.)
+						$block_class_names = explode( ' ', $block_attrs['className'] );
```
I suppose that every block uses the space character (U+20) as a class name separator, but should we match HTML’s behavior here and use `" \t\f\r\n"`? also, should we call `array_unique()` on it? I think the use of `class_list()` removes duplicates, but they are left in here for the block’s classes

ockham replied:
Good point! Done in https://github.com/WordPress/wordpress-develop/pull/8882/commits/bb2404b77b0d9cad599459a3b705c774681e2cc2 and https://github.com/WordPress/wordpress-develop/pull/8882/commits/192009bd006705bbd8fa0af77ed77de26f624665.

ockham committed:
Use WP_Block_Parser for block parsing

ockham committed:
Handle void blocks correctly

ockham committed:
Rearrange logic

ockham committed:
Allow for different whitespace in class attribute

ockham committed:
De-duplicate block class names

dmsnell commented on tests/phpunit/includes/build-tree-representation.php:
```diff
@@ -0,0 +1,267 @@
+							$block_class_names = preg_split( '/[ \t\f\r\n]+/', trim( $block_attrs['className'], " \f\t\r\n" ) );
+							$block_class_names = array_unique( $block_class_names, SORT_STRING );
+							sort( $block_class_names, SORT_STRING );
+							$block_attrs['className'] = implode( ' ', $block_class_names );
```
huh. well you know what, we’re just duplicating logic from the Tag Processor. 🤔 
perhaps we ask it to do this?

```php
$block_class_processor = new WP_HTML_Tag_Processor( '<div>' );
$block_class_processor->next_token();
$block_class_processor->set_attribute( 'class', $block_attrs['className'] );
$block_attrs['className'] = iterator_to_array( $block_class_processor->class_list() );
sort( $block_attrs['className'] );
$block_attrs['className'] = implode( ' ', $block_attrs['className'] );
```

it’s perhaps not less code, but it _is_ leaving the CSS class name handling to the tested code that does that

ockham replied:
Good idea! Done in 252569dce8.

ockham committed:
Change Html5lib test format reference link

ockham committed:
Add more PHPDoc and an example to build_tree_representation

ockham committed:
Use Tag Processor to normalize block class names

ockham committed:
Whitespace :rolleyes:

ockham committed:
Rename function to build_equivalent_html_semantic_tree

ockham committed:
Add covers PHPDoc to tests

ockham committed:
Fall back to DOM\HTMLDocument for PHP 8.4+

ockham committed:
Revert "Bring back assertEqualMarkup override, harmonize signature"

This reverts commit ef27f4fb3f6d6469d806c91be40d8e16cba41676.

ockham committed:
Rename new method to assertEqualBlockMarkup

ockham committed:
Add @todo note to "old" assertEqualMarkup

ockham committed:
Update PHPDoc for assertEqualBlockMarkup

aaronjorbin commented on tests/phpunit/includes/abstract-testcase.php:
```diff
@@ -1180,6 +1180,44 @@ public function assertQueryTrue( ...$prop ) {
+	 *                                      Set to null to parse a full HTML document.
+	 * @param string|null $message          Optional. The assertion error message.
+	 */
+	public function assertEqualBlockMarkup( string $expected, string $actual, ?string $fragment_context = '<body>', $message = 'HTML markup was not equivalent.' ): void {
```
If I'm understanding this correctly, it isn't just Block's that this could be useful for, it's any HTML.  I wonder if a name like `assertEqualHTML` would be better to make that clearer

ockham replied:
I've discussed this with @gziolo today, and he agrees. We also have @sirreal's 👍 emoji reaction, so:

Renamed in b07677e472970956b3262866e2f835a3bbb975bc.

aaronjorbin commented on tests/phpunit/tests/build-equivalent-html-semantic-tree.php:
```diff
@@ -0,0 +1,134 @@
+			'Duplicated block class names'             => array(
+				'<!-- wp:separator {"className":"wp-block-separator is-style-default"} -->',
+				'<!-- wp:separator {"className":"wp-block-separator is-style-default wp-block-separator"} -->',
+			),
```
I thiink it would also be helpful to include a test with capitlization/lower case letters. 
```suggestion
			),
			'Different Capitalization of tag'             => array(
				'<IMG src="wp.png" alt="The WordPress logo">',
				'<img src="wp.png" alt="The WordPress logo">',		
			),	
```

ockham replied:
Good idea 👍

aaronjorbin reviewed (commented):
Love the idea. Have you tried testing to see if this can be a replacement for `assertEqualMarkup`?

ockham committed:
Add test with capitalization/lower case letters.

Co-authored-by: Aaron Jorbin <aaronjorbin@users.noreply.github.com>

ockham commented:
> Love the idea. Have you tried testing to see if this can be a replacement for `assertEqualMarkup`?

I would like it to be eventually, but I couldn't use it as a drop-in replacement. (There's more on that in the PR description, under "Note".) As to the nature of the discrepancies: AFAIR, it was mostly that `assertEqualMarkup` removed `<[CDATA` while `assertEqualBlockMarkup` (i.e. the HTML API) kept it; plus boolean attributes such as `async` and `defer` on `<script>` tag. (IIRC, WP's own util functions -- i.e. enqueue/register script -- use the old `async="async"` notation, whereas the HTML API uses the HTML5 style -- i.e. just `async`.)

My takeaway was that it would mean a bit more work to sort out the ~50 (?) test failures I was getting when I tried this, and I didn't want to block the PR by that. Eventually, I'm all for retiring `scripts.php`'s `assertEqualMarkup` in favor of an HTML API-based assertion, though.

sirreal committed:
Update to use the new assertEqualMarkup helper

sirreal committed:
Rename test helper to assertEqualMarkup

sirreal committed:
Fix script async/defer strategy expectation to match printed scripts

sirreal committed:
Add CDATA inline script wrappers where they appear in markup

sirreal committed:
FIXME: Skip an inaccurate test

sirreal committed:
Adjust expected output to use attributes as generated

The generated HTML produces async="async" or defer="defer".
Match this in the expected HTML.

sirreal committed:
Add html5 support to a test that expects html5 support

sirreal committed:
Script quotes in IE7 conditional _comments_ produce semantically _different_ HTML

sirreal committed:
Revert irrelevant change

sirreal committed:
Fix more missing CDATA comments

sirreal committed:
Remove scripts specific assertEqualMarkup implementation

sirreal committed:
Use more consistent code backticks in docblock

sirreal commented:
I've pushed some changes to completely remove the previous `assertEqualMarkup` with this implementation.

The work discovered at least one bug in the tests where the previous implementation "normalized" the output and forced the markup to _appear_ equivalent when, in face, it was not. In this case, it purported to test html5 script behavior, but did not set html5 theme support.

https://github.com/WordPress/wordpress-develop/blob/b0df96a0d6cec2bceb5b4e16817f71b05fee70e0/tests/phpunit/tests/dependencies/scripts.php#L1531-L1534

There are a number of changes where `/* <![CDATA[ */ /* ]]> */` wrappers are introduced around script tag contents or attributes were produced like `defer="defer"` instead of booleans.

It's clear that this implementation of `assertEqualMarkup` provides a much more accurate representation of the markup.

---

There's a conversation to be had around [HTML5 support](https://core.trac.wordpress.org/ticket/59883) and whether the non HTML5 behaviors even make sense today, but that's beyond the scope of this PR.

sirreal committed:
Fix phpcs issues

sirreal commented on tests/phpunit/tests/dependencies/scripts.php:
```diff
@@ -2023,8 +2141,8 @@ public function test_wp_add_inline_script_after_with_concat_and_conditional_and_
 		$expected .= "<!--[if gte IE 9]>\n";
-		$expected .= "<script type='text/javascript' src='http://example.com' id='test-example-js'></script>\n";
-		$expected .= "<script type='text/javascript' id='test-example-js-after'>\nconsole.log(\"after\");\n</script>\n";
+		$expected .= "<script type=\"text/javascript\" src=\"http://example.com\" id=\"test-example-js\"></script>\n";
+		$expected .= "<script type=\"text/javascript\" id=\"test-example-js-after\">\n/* <![CDATA[ */\nconsole.log(\"after\");\n/* ]]> */\n</script>\n";
```
There are a few cases like this where the contents of an (IE conditional) _HTML comment_ were changed.

[These script tags, on their own, would be recognized as equivalent](https://software.hixie.ch/utilities/js/live-dom-viewer/?%3C!DOCTYPE%20html%3E%0A%3Cscript%20type%3D%27text%2Fjavascript%27%3E%3C%2Fscript%3E%0A%3Cscript%20type%3D%22text%2Fjavascript%22%3E%3C%2Fscript%3E) with double `"` or single `'` quotes. But they're not script tags, [they are the text contents of an HTML comment ](https://software.hixie.ch/utilities/js/live-dom-viewer/?%3C!DOCTYPE%20html%3E%0A%3C!--%5Bif%20gte%20IE%209%5D%3E%0A%3Cscript%20type%3D%27text%2Fjavascript%27%3E%3C%2Fscript%3E%0A%3C!%5Bendif%5D--%3E%0A%3C!--%5Bif%20gte%20IE%209%5D%3E%0A%3Cscript%20type%3D%22text%2Fjavascript%22%3E%3C%2Fscript%3E%0A%3C!%5Bendif%5D--%3E)where `'` is not the same as `"`.

sirreal commented:
All of the tests here for IE conditional comments also seem quite out of place in today's internet.

ockham committed:
Remove now-obsolete parse_markup_fragment() method

sirreal commented on tests/phpunit/includes/build-equivalent-html-semantic-tree.php:
```diff
@@ -0,0 +1,299 @@
+ * @param string      $html             Given test HTML.
+ * @return string|null Tree structure of parsed HTML, if supported, else null.
+ */
+function build_equivalent_html_semantic_tree( string $html, ?string $fragment_context ) {
```
There's no equivalence at this point. It's a single tree representation of the HTML (and block) semantics. Equivalence only enters the picture in the assertion where we compare two trees to understand whether they're semantically equivalent.

Let's remove the notion of equivalence from this function name:

```suggestion
function build_html_semantic_tree( string $html, ?string $fragment_context ) {
```

dmsnell replied:
while we’re at it, it seems like the _semantic_ part is stored inside the HTML API and not in the output text. is this actually more a _visual_ tree?

```php
function build_visual_html_tree() {}
```

I ask because if we wanted to compare apples to apples I think we can just normalize two documents, or provide two HTML Processors, but I think a reasonable part of this function is providing better error display when the HTML is not equivalent?

sirreal replied:
That's a good point, I've struggled to figure out the language myself. It's a _visual_ or human-readable representation of the tree. I landed on semantic because, although it's imperfect, it conveys that it represents the semantics of the HTML rather than the syntax.

It may be the case that "tree" already suggests that it's a processed form of the HTML input.

I'm happy with "visual," more details about what the function does are available in the PHP doc comment.

ockham replied:
Alright, I'll rename to `build_visual_html_tree`!

ockham replied:
9469f43d0bd85ffbc7c6ea24f9db3d724d6a9f3f

sirreal commented on tests/phpunit/includes/build-equivalent-html-semantic-tree.php:
```diff
@@ -0,0 +1,299 @@
+ *
+ * That format is extended with a special representation of block delimiters and their attributes.
+ * Furthermore, it the order of attributes and class names is normalized both for HTML tags and block delimiters,
+ * as is the whitespace in HTML tags' style attribute.
```
Some suggestions to:

- Make the function short description more clear and to the point. html5lib isn't relevant to the essential behavior of this function.
- Improve some language in the description around blocks and talk about block semantics instead of delimiters.

These lines may need to be properly formatted.

```suggestion
 * Generates representation the semantic HTML tree structure. 
 *
 * This is inspired by the representation used by the HTML5lib tests. It's been extended here for 
 * blocks to render the semantic structure of blocks and their attributes.
 * The order of attributes and class names is normalized both for HTML tags and blocks,
 * as is the whitespace in HTML tags' style attribute.
```

sirreal commented on tests/phpunit/includes/build-equivalent-html-semantic-tree.php:
```diff
@@ -0,0 +1,299 @@
+ * @see https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/data/html5lib-tests/tree-construction/README.md
+ *
+ * @param string|null $fragment_context Context element in which to parse HTML, such as BODY or SVG.
+ * @param string      $html             Given test HTML.
```
These got out of order.
```suggestion
 * @param string      $html             Given test HTML.
 * @param string|null $fragment_context Context element in which to parse HTML, such as BODY or SVG.
```

sirreal commented on tests/phpunit/includes/build-equivalent-html-semantic-tree.php:
```diff
@@ -0,0 +1,299 @@
+ * @param string      $html             Given test HTML.
+ * @return string|null Tree structure of parsed HTML, if supported, else null.
+ */
+function build_equivalent_html_semantic_tree( string $html, ?string $fragment_context ) {
```
We can add a return type. On review, it appears this returns a string or throws:

```suggestion
 * @return stringTree structure of parsed HTML, if supported, else null.
 */
function build_equivalent_html_semantic_tree( string $html, ?string $fragment_context ): string {
```

It's probably a good idea to annotate `@throws …`.

ockham replied:
Done in 6a7921663c.

sirreal reviewed (approved):
I've left some feedback but I think this is great and ready to go for the most part. It becomes much easier and less error prone to compare HTML.

It's worth mentioning that the tree builder functionality is largely duplicated from existing html5lib tests used for the HTML API. It's already seen significant usage, although the block parsing is new.

https://github.com/WordPress/wordpress-develop/blob/75f4a0951f0c75a7868274570807975465684923/tests/phpunit/tests/html-api/wpHtmlProcessorHtml5lib.php#L159

ockham committed:
Fix PHPDoc param order

Co-authored-by: Jon Surrell <sirreal@users.noreply.github.com>

ockham committed:
Tweak PHPDoc description of build_equivalent_semantic_tree

Co-authored-by: Jon Surrell <sirreal@users.noreply.github.com>

ockham committed:
Add return type, tweak PHPDoc

ockham commented:
> It's worth mentioning that the tree builder functionality is largely duplicated from existing html5lib tests used for the HTML API. It's already seen significant usage, although the block parsing is new.

@sirreal Think it's worth doing a follow-up to share code between the two?

ockham committed:
Rename to build_visual_html_tree

ockham committed:
Rename to assertEqualHTML

sirreal commented:
I only mentioned it because it seemed relevant for other reviewers. We could try out the sharing but I don't think it's necessary.

ockham committed:
Fix indentation

ockham committed:
Add ticket annotations
